Hoist CI build requirements to top of issue template

jtigger opened the pull request:
Feels like a lot of unilateral decisions in this PR; I hope I'm understanding the purpose and audience.

1. suggesting that we make explicit all the "requirements" of setting up CI at the very top of the file (the what).
2. elaborate the background on each "requirement" with similarly named sections (the why).
3. keep the excellent examples at the tail of the file (the how).

Also, there is background for writing high quality exercise test suites.  Suggesting we extract these to a separate file.  In fact, I believe this would go in the new docs repo in some section giving background on writing test suites.

kytrinyx commented:
> In fact, I believe this would go in the new docs repo in some section giving background on writing test suites.

Yes, I think that is right. I think that probably goes in `contributing-to-language-tracks`, but `maintaining-a-track` would also be correct.

kytrinyx commented on TRAVIS:
```diff
@@ -1,6 +1,13 @@
+
+- for each exercise in the track,
+  - run that exercise's test suite (aka the "exercise test suite");
+  - use the exercise's solution as the code under test.
```
"sample solution" or "example solution"?

jtigger replied:
👍 the tooling is looking for an "example" solution, that one.

kytrinyx commented on TRAVIS:
```diff
@@ -1,6 +1,13 @@
+- for each exercise in the track,
+  - run that exercise's test suite (aka the "exercise test suite");
+  - use the exercise's solution as the code under test.
+- ensure all test cases are enabled;
```
Should ensuring all test cases are enabled be first in the list?

jtigger replied:
Noice.  In fact, I think it would best serve as a sub-bullet of "run that exercise's test suite".

kytrinyx commented on TRAVIS:
```diff
@@ -1,6 +1,13 @@
+- ensure all test cases are enabled;
+- be executed on Travis CI;
+- be easy to run on a contributor/maintainer's local machine (i.e. a build script with any needed instructions).
 
```
Ah, yes. Instructions. Let's specify that this needs to be documented in the README of the track.

parkerl replied:
There is a placeholder for this as the last bullet under "Submitting a Pull Request" in the TRACK_README template.

jtigger replied:
@parkerl would that bullet _contain_ the instructions to run the script?  Or would it reference those instructions?  (on some tracks, there might be a little more setup surrounding the script execution... *...cough...* Java *...cough...* for example *...wheeez...*).

kytrinyx commented on TRAVIS:
```diff
@@ -9,34 +16,31 @@ Implement a track test suite that can run both locally and on Travis CI. The tra
+_Enabling all test cases_
 
-Many test frameworks will randomize the order of the tests when running them. This is an excellent practice, which helps ensure that subsequent tests are not dependent on side effects from earlier tests. However, in order to simulate TDD we want tests to run *in the order that they are defined*, and we want them to *fail fast*, that is to say, as soon as the test suite encounters a failure, we wa ...
+As described in [Contributing to Exercism: Writing exercise test suites](writing_tests.md), exercise test suites are written such that all but the first test case are skipped/pending/ignored.  For the purpose of the track test suite, all test cases should be enabled (simulating an Exercism user having completed the exercise).
```
- `s/Writing/writing/`
- `s/writing_tests.md/writing-tests.md`

> exercise test suites are written such that all but the first test case are skipped/pending/ignored.

`s/test suites/test suites in most language tracks/`. Some language tracks are able to implement this without skipping, pending, ignoring (e.g. Go has a `t.Fatal()` that will stop on the first error or failure, whereas `t.Error()` will report the error and move on. I think also some language tracks have decided not to fail fast. That might be a decision that we need to consider when we start looking at the curriculum from a high level and consider the user experience.

I really like the _simulating an Exercism user having completed the exercise_ bit.

parkerl replied:
👍

kytrinyx commented on TRAVIS:
```diff
@@ -9,34 +16,31 @@ Implement a track test suite that can run both locally and on Travis CI. The tra
+We will need to temporarily (and programmatically) edit the exercise test suites to ensure that all of their tests are active. We may also need to rename the example solution file(s) in order for the exercise test suite to run against it.
 
-Whatever the mechanism—functions, methods, annotations, directives, commenting out tests, or some other approach—these are changes made directly to the test file. The person solving the exercise will need to edit the test file in order to "activate" each subsequent test.
+It's important that if we rewrite files in any way during a test run, that these changes do not accidentally get checked in to the git repository.
```
`s/git/Git/` (Git is a proper name)

kytrinyx commented on TRAVIS:
```diff
@@ -9,34 +16,31 @@ Implement a track test suite that can run both locally and on Travis CI. The tra
+It's important that if we rewrite files in any way during a test run, that these changes do not accidentally get checked in to the git repository.
 
-Any tests that are marked as skipped will not be verified by the track test suite unless special care is taken.
+Therefore, many language tracks write the track test suite in such a way that it _copies_ the exercise to a temporary location outside of the git repository before editing or rewriting the exercise files during a test run.
```
`s/git/Git`

kytrinyx commented on TRAVIS:
```diff
@@ -9,34 +16,31 @@ Implement a track test suite that can run both locally and on Travis CI. The tra
+Therefore, many language tracks write the track test suite in such a way that it _copies_ the exercise to a temporary location outside of the git repository before editing or rewriting the exercise files during a test run.
 
-Additionally, in some programming languages, the name of the file containing the solution is hard-coded in the test suite, and the example solution is not named in the way that we expect people to name their files.
+_Configuring Travis CI_
```
I now realize that we sometimes have AppVeyor (Microsoft stack) and Circle CI (Apple stack). Travis CI only runs on Linux.

kytrinyx commented on TRAVIS:
```diff
@@ -9,34 +16,31 @@ Implement a track test suite that can run both locally and on Travis CI. The tra
-We will need to temporarily (and programmatically) edit the exercise test suites to ensure that all of their tests are active. We may also need to rename the example solution file(s) in order for the exercise test suite to run against it.
+A contributor may forget to run the track test suite before submitting a PR.  Also, sometimes a test suite will succeed even when something is wrong with the test suite on a contributor's local machine because of some artifact of their local configuration.  Running the suite on a continuous integration server (here, Travis CI) ensures the track test suite is run from scratch.
 
-**Avoiding accidental git check-ins**
```
And I clearly forgot that Git is a proper name. Shame on me.

parkerl replied:
🤣

kytrinyx commented on writing_tests.md:
```diff
@@ -0,0 +1,16 @@
+_(in fact, this would be part of the new "docs" repo, not a file in this repo.)_
```
Agreed.

parkerl replied:
Part of this was stolen by "someone" and placed in the TRACK_README. I agree a part of it should also go into docs but some of this about randomization is relevant here.

kytrinyx replied:
Yeah, how about a one-liner that links to the docs?

jtigger replied:
... Also, I move to hunt down that "someone" and bring him to justice! ;) 🍻

jtigger replied:
_(clears his throat to appear serious)_ So, which parts do you feel should be in-lined, @parkerl ?  As I'm suggesting in this PR, we're really nudging pretty hard to go read that section in the central guide.  I'm of the mind that these characteristics (including how some tooling randomizes tests while others don't) are really part and parcel to this general discussion about test suites.

jtigger commented:
Okay, I'm really diggin' the copy editing!  Thank you for the thorough review, folks.

I believe I addressed the points brought up.  A little discussion still about landing spots for a few bits.

kytrinyx commented:
> A little discussion still about landing spots for a few bits.

Would you rephrase? I can't figure out how to parse this. (Sorry! I am traveling, brain might not actually be working right) 🌷

jtigger commented:
Sorry for the delay, @kytrinyx.  In fact, I missed your suggestion to put in a one-liner and link to the docs.  That's a perfect solution, here.

That reference would be to the page requested by exercism/docs#8.  I'm gonna update that issue to request the reference be made as part of that work.

parkerl commented:
@jtigger I'm lost. Where are we with this one?

kytrinyx commented:
I ended up removing the issue template, as it contained too many assumptions about the state of the track. I'm going to go ahead and close this issue.
